src/net_processing.cpp

```diff
@@ -144,6 +144,21 @@ void static ProcessGetData(CNode *pfrom, const Consensus::Params &consensusParam
                     pfrom->fDisconnect = true;
                     fSend = false;
                 }
+                // Avoid leaking prune-height by never sending blocks below the
+                // NODE_NETWORK_LIMITED threshold.
+                // Add two blocks buffer extension for possible races
+                if (send && !pfrom->fWhitelisted &&
+                    ((((nLocalServices & NODE_NETWORK_LIMITED) == NODE_NETWORK_LIMITED) &&
+                        ((nLocalServices & NODE_NETWORK) != NODE_NETWORK) &&
+                        (chainActive.Tip()->nHeight - mi->nHeight > (int)NODE_NETWORK_LIMITED_MIN_BLOCKS + 2))))
```

Comment:
`chainActive.Tip()` needs `cs_main`

Reply:
This whole if clause could be extracted into a function, for example `can_network_limited_send` for readability

Reply:
we've removed the requirement that Tip() hold cs_main.